sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarPartitionReaderWithPartitionValues.scala

```diff
@@ -114,4 +178,70 @@ object ColumnarPartitionReaderWithPartitionValues {
       }
     }
   }
+
+  // scalastyle:off line.size.limit
+  /**
+   * Splits the partition into smaller batches such that the batch size
+   * does not exceed cudfLimit for any column.
+   *
+   * Algorithm:
+   * - Calculate the max size of single value among partition columns.
+   * - Calculate the size of partition = numRows * maxSizeOfSingleValue.
+   * - If size of partition exceeds cudfLimit, split the partition into batches.
+   *    - Calculate newRows = cudfLimit / maxSizeOfSingleValue.
+   *    - Add newRows to resultBatches.
+   *    - Update remainingNumRows and remainingSizeOfPartition.
+   *    - Repeat until remainingSizeOfPartition is less than cudfLimit.
+   *  - Return the array of batch sizes in resultBatches.
+   *
+   * Example:
+   * Input:
+   *    numRows: 50
+   *    partValues: [ abc, abcdef ]
+   *    cudfLimit: 140 bytes
+   *
+   * Result:
+   *    [ 23, 23, 4 ]
+   *
+   * @return An array of batch sizes, such that the size of each batch is
+   *         below cudfLimit.
+   */
+  // scalastyle:on line.size.limit
+  private def splitPartitionIntoBatchesScalar(numRows: Int,
+      partValues: Array[Scalar], sparkTypes: Array[DataType]): Array[Int] = {
+    val cuDFLimit: Long = (1L << 31) - 1
+    // Calculate the maximum size of a String type column.
+    // Assumption: Columns of other data type would not have fit already.
+    val maxSizeOfSingleValue: Int = sparkTypes.zipWithIndex.collect {
+      case (StringType, colIndex) => partValues(colIndex).getUTF8.length
+    }.max
```

Comment:
This will crash if no values were collected because there were no string partition columns